File: src/features/region-selector/regionSelectionTypes.ts
import {
  PERFORMANCE_LIMITS,
  type RegionSize
} from "../performance/performanceLimits";

const MIN_REGION_SIZE: RegionSize = {
  width: 24,
  height: 24
};

export interface LogicalPoint {
  x: number;
  y: number;
}

export interface PhysicalPoint {
  x: number;
  y: number;
}

export interface MonitorGeometry {
  id: string;
  logicalOrigin: LogicalPoint;
  physicalOrigin: PhysicalPoint;
  scaleFactor: number;
}

export interface RegionSelectionRequest {
  monitor: MonitorGeometry;
  start: LogicalPoint;
  end: LogicalPoint;
}

export interface PhysicalRegion {
  monitorId: string;
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface RegionSelectionLimits {
  minimumRegion: RegionSize;
  recommendedRegion: RegionSize;
  maxRegion: RegionSize;
}

export interface RegionSelection {
  region: PhysicalRegion;
  warnings: RegionSelectionIssue[];
  limits: RegionSelectionLimits;
}

export type RegionSelectionIssueCode =
  | "invalidCoordinate"
  | "invalidScaleFactor"
  | "regionTooNarrow"
  | "regionTooShort"
  | "regionCoordinateOutOfRange"
  | "regionWidthAboveRecommended"
  | "regionHeightAboveRecommended"
  | "regionWidthTooLarge"
  | "regionHeightTooLarge";

export interface RegionSelectionIssue {
  code: RegionSelectionIssueCode;
  limit: number;
  actual: number;
  message: string;
}

export type RegionSelectionResult =
  | { ok: true; selection: RegionSelection }
  | { ok: false; error: RegionSelectionIssue };

export const REGION_SELECTION_LIMITS: RegionSelectionLimits = {
  minimumRegion: MIN_REGION_SIZE,
  recommendedRegion: PERFORMANCE_LIMITS.recommendedRegion,
  maxRegion: PERFORMANCE_LIMITS.maxRegion
};

export const ISSUE_MESSAGES: Record<RegionSelectionIssueCode, string> = {
  invalidCoordinate: "Selection coordinates must be finite.",
  invalidScaleFactor: "Monitor scale factor must be greater than zero.",
  regionTooNarrow: "Selected region is too narrow.",
  regionTooShort: "Selected region is too short.",
  regionCoordinateOutOfRange:
    "Selected region is outside the supported coordinate range.",
  regionWidthAboveRecommended: "Selected region is wider than recommended.",
  regionHeightAboveRecommended: "Selected region is taller than recommended.",
  regionWidthTooLarge: "Selected region is wider than allowed.",
  regionHeightTooLarge: "Selected region is taller than allowed."
};

File: src-tauri/src/region_selection_types.rs
use serde::{Deserialize, Serialize};

use crate::performance_limits::RegionSize;

#[cfg(target_os = "macos")]
use std::sync::Arc;

#[cfg(target_os = "macos")]
use screencapturekit::shareable_content::SCWindow;

#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LogicalPoint {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LogicalSize {
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PhysicalPoint {
    pub x: i32,
    pub y: i32,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MonitorGeometry {
    pub id: String,
    pub logical_origin: LogicalPoint,
    pub logical_size: LogicalSize,
    pub physical_origin: PhysicalPoint,
    pub scale_factor: f64,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegionSelectionRequest {
    pub monitor: MonitorGeometry,
    pub start: LogicalPoint,
    pub end: LogicalPoint,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WindowCaptureTarget {
    pub window_id: u32,
    pub owner_pid: i32,
    pub source_width_millipoints: u64,
    pub source_height_millipoints: u64,
    #[cfg(target_os = "macos")]
    pub native_window: Option<Arc<SCWindow>>,
    pub relative_x_millipoints: i64,
    pub relative_y_millipoints: i64,
    pub width_millipoints: u64,
    pub height_millipoints: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PhysicalRegion {
    pub monitor_id: String,
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
    #[serde(skip)]
    pub source_window: Option<WindowCaptureTarget>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegionSelectionLimits {
    pub minimum_region: RegionSize,
    pub recommended_region: RegionSize,
    pub max_region: RegionSize,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegionSelection {
    pub region: PhysicalRegion,
    pub warnings: Vec<RegionSelectionIssue>,
    pub limits: RegionSelectionLimits,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegionSelectionIssue {
    pub code: RegionSelectionIssueCode,
    pub limit: f64,
    pub actual: f64,
    pub message: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum RegionSelectionIssueCode {
    InvalidCoordinate,
    InvalidScaleFactor,
    SelectionOutsideMonitor,
    RegionTooNarrow,
    RegionTooShort,
    RegionCoordinateOutOfRange,
}

pub type RegionSelectionResult = Result<RegionSelection, RegionSelectionIssue>;
